src/Microsoft.Health.Fhir.Core/Features/Search/Hackathon/QueryPlanSelector.cs

```diff
@@ -0,0 +1,108 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using EnsureThat;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Health.Fhir.Core.Features.Search.Hackathon
+{
+    public sealed class QueryPlanSelector : IQueryPlanSelector<bool>
+    {
+        private const int MaxEwmaEntries = 500;
+        private const int MinIterationsForDecision = 6;
+        private const double EwmaAlpha = 0.3;
+
+        // Possible settings for enabling or disabling query plan caching.
+        // 'False' is the first element to ensure it is the default when EWMA scores are equal.
+        private readonly string[] _settings = new[] { "False", "True" };
+
+        private readonly object _lock;
+        private readonly Dictionary<string, Ewma> _ewmaByHash;
+        private readonly ILogger<QueryPlanSelector> _logger;
+
+        public QueryPlanSelector(ILogger<QueryPlanSelector> logger)
+        {
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _lock = new object();
+            _ewmaByHash = new Dictionary<string, Ewma>();
+        }
+
+        public bool GetQueryPlanCachingSetting(string hash)
+        {
+            EnsureArg.IsNotNullOrEmpty(hash, nameof(hash));
+
+            Ewma ewma;
+            lock (_lock)
+            {
+                ewma = GetEwma(hash);
+            }
+
+            // Hard max limit to avoid memory leaks while using dictionaries.
+            // TODO: use dotnet memory cache with expiration instead of limiting the size of the dictionary.
+            if (ewma == null)
+            {
+                if (IsMaxLimitReached())
+                {
+                    return false;
```

Comment:
This needs a way to set the default response based on the server setting. False isn't always the default.